File: python/aoc.py
#!/usr/bin/env python
#
# usage: aoc.py [-y|--year year] [-d|--day day] [-t|--test] 1|2

import argparse
import datetime
import importlib
import os.path
import sys

import .testing

now = datetime.datetime.today()
year = now.year
day = now.day

parser = argparse.ArgumentParser()
parser.add_argument("-y", "--year", type=int, help="year")
parser.add_argument("-d", "--day", type=int, help="day")
parser.add_argument("-t", "--test", action="store_true", help="test mode")
parser.add_argument("part_number", type=int, help="puzzle part number (1 or 2)")
args = parser.parse_args()
if args.year:
    year = int(args.year)
if args.day:
    day = int(args.day)

module = importlib.import_module(f"y{year}.day{'%02d' % day}")

# NOTE: this will only work for years 2025+ because in the last few years
# I've drastically changed how I run tests. See ../ruby. Previous years'
# solutions won't work with run_chunk_tests, they have the testing code
# built in. This code won't handle that.
func = getattr(module, f"part{int(args.part_number)}", None)
if args.test:
    testing.run_chunk_tests(args, lambda expected, lines: expected == func(ctx, lines))
else:
    print(func(args))

File: ruby/aoc.rb
#!/usr/bin/env ruby
#
# usage: aoc.rb [-y|--year=year] [-d|--day-day] [--D|--DEBUG] [-t|--test] 1|2

require 'optparse'
require_relative 'day'

# If we pass in a file name, we parse the year and day of the month from its
# path. Else we start with the current year and day of the month. Both can
# be overridden by command line arguments.
def aoc(file = nil)
  if file
    day = File.basename(file)[3, 2].to_i
    year = File.basename(File.dirname(file))[1..].to_i
  else
    now = Time.now
    day = now.day
    year = now.year
  end
  testing = false

  OptionParser.new do |opts|
    opts.on('-y YEAR', '--year YEAR', 'year') { |arg| year = arg.to_i }
    opts.on('-d DAY', '--day DAY', 'day') { |arg| day = arg.to_i }
    opts.on('-t', '--testing', 'testing') { |_| testing = true }
    opts.on('-D', '--debug', 'debug') { |_| $DEBUG = true }
  end.parse!
  if ARGV[0].nil?
    warn 'error: must specify part number 1 or 2'
    exit(1)
  end
  part_number = ARGV[0].to_i
  day_str = '%02d' % day

  klass_name = "Day#{day_str}"
  require_relative "y#{year}/day#{day_str}.rb" unless Object.const_defined?(klass_name)
  Object.const_get(klass_name).new(year, day, part_number, testing).run
end

aoc if __FILE__ == $PROGRAM_NAME
